File: online-training-platform/leetcode/backspace-string-compare.cpp
class Solution {
public:
    
    string realString(string str){
        int s = str.size();
        string r = "";
        int c = 0, j = 0, m = 0, l = 0,k = 0;
        for(int i = s-1 ; i >= 0 ;){
            if(str[i] != '#'){
                r= str[i] + r;
                i--;
            }else{
                
                if(i == 0) break;
                j = i;
                m = 0;
                l = 0;
                while(str[j] == '#'){
                    m++;
                    j--;
                }
                k = j;
                while( k >= 0 && str[k] != '#' ){
                    l++;
                    k--;
                }
                
                if( l > m ){
                    j -= m;
                    k = l - m;
                    while(c > 0 && k > 0 && j >= 0){
                        c--;
                        j--;
                        k--;
                    }    
                }else{
                    j -= l;
                    c += (m-l);    
                }
                i = j;
            }
        }
        return r;
    }
    
    bool backspaceCompare(string S, string T) {
        return realString(S) == realString(T);
    }
};

File: online-training-platform/leetcode/backspace-string-compare.kt
class Solution {
    
    fun getString(str: String): String{
        val sb = StringBuilder()
        str.forEachIndexed{ idx, c ->
            val sz = sb.length
            if(c == '#' && sz > 0){
                sb.setLength(sz - 1)
            } else if(c != '#'){
                sb.append(c)
            }
        }
        
        return sb.toString()
    }
    
    fun backspaceCompare(s: String, t: String): Boolean {
        
        val ss = getString(s)
        val st = getString(t)
        
        println(ss)
        println(st)
        
        return ss == st
    }
}